net/src/ipv4/mod.rs

```diff
@@ -31,10 +31,16 @@ impl Ipv4 {
     #[allow(unsafe_code)] // const-eval and trivially safe
     pub const MAX_LEN: NonZero<usize> = unsafe { NonZero::new_unchecked(60) };
 
+    /// Create a new IPv4 header
+    pub(crate) fn new(header: Ipv4Header) -> Result<Self, Ipv4Error> {
+        UnicastIpv4Addr::new(Ipv4Addr::from(header.source)).map_err(Ipv4Error::InvalidSourceAddr)?;
+        Ok(Self(header))
+    }
+
     /// Get the source ip address of the header
     #[must_use]
-    pub fn source(&self) -> Ipv4Addr {
-        Ipv4Addr::from(self.0.source)
+    pub fn source(&self) -> UnicastIpv4Addr {
```

Comment:
I love this change.  I should have written it this way from the start.

In a perfect world we would be able to do things more like TypeScript + [Zod's brand method](https://github.com/colinhacks/zod?tab=readme-ov-file#brand) (oh the irony) to eliminate commutativity and associativity issues with qualifier types like this (while also avoiding an infinity mirror of increasingly complex types).

In the mean time, I think this is a great change.